File: Tarea4/Python/Tarea04.py
import sounddevice as sd
import numpy as np
import wave
import matplotlib.pyplot as plt
from scipy.io import wavfile
from scipy.signal import welch

def grabar_audio(duracion, archivo):
    fs = 44100  # Frecuencia de muestreo
    print("Comenzando la grabación...")
    audio = sd.rec(int(duracion * fs), samplerate=fs, channels=1, dtype='float32')
    sd.wait()  # Espera a que la grabación termine
    print("Grabación finalizada.")
    # Guardar el archivo de audio
    audio = np.array(audio)
    audio = np.squeeze(audio)
    with wave.open(archivo, 'w') as wf:
        wf.setnchannels(1)  # Canal mono
        wf.setsampwidth(2)  # 16 bits por muestra
        wf.setframerate(fs)
        wf.writeframes((audio * 32767).astype(np.int16).tobytes())
    print(f"Archivo de audio grabado correctamente como {archivo}.")

def reproducir_audio(archivo):
    try:
        # Leer archivo WAV
        fs, data = wavfile.read(archivo)
        sd.play(data, fs)
        sd.wait()  # Espera a que la reproducción termine
    except Exception as e:
        print(f"Error al reproducir el audio: {e}")

def graficar_audio(archivo):
    try:
        fs, data = wavfile.read(archivo)
        tiempo = np.linspace(0, len(data) / fs, num=len(data))
        plt.figure()
        plt.plot(tiempo, data)
        plt.xlabel('Tiempo (s)')
        plt.ylabel('Amplitud')
        plt.title('Audio')
        plt.show()
    except Exception as e:
        print(f"Error al graficar el audio: {e}")

def graficar_densidad(archivo):
    try:
        fs, data = wavfile.read(archivo)
        f, Pxx = welch(data, fs, window='hann', nperseg=len(data))
        plt.figure()
        plt.semilogy(f, Pxx)
        plt.xlabel('Frecuencia (Hz)')
        plt.ylabel('Densidad espectral de potencia (dB/Hz)')
        plt.title('Espectro de frecuencia del audio grabado')
        plt.show()
    except Exception as e:
        print(f"Error al graficar la densidad espectral: {e}")

def menu():
    opcion = 0
    archivo = 'audio.wav'
    while opcion != 5:
        print("Seleccione una opción:")
        print("1. Grabar")
        print("2. Reproducir")
        print("3. Graficar")
        print("4. Graficar densidad")
        print("5. Salir")
        try:
            opcion = int(input("Ingrese su elección: "))
            if opcion == 1:
                duracion = int(input("Ingrese la duración de la grabación en segundos: "))
                grabar_audio(duracion, archivo)
            elif opcion == 2:
                reproducir_audio(archivo)
            elif opcion == 3:
                graficar_audio(archivo)
            elif opcion == 4:
                graficar_densidad(archivo)
            elif opcion == 5:
                print("Saliendo del programa...")
            else:
                print("Opción no válida.")
        except ValueError:
            print("Opción no válida.")

if __name__ == "__main__":
    menu()

File: Tarea4/Octave/Tarea04.m
% Comprueba si estamos ejecutando en MATLAB o en Octave
if (exist('OCTAVE_VERSION', 'builtin') ~= 0)
% Estamos en Octave
  pkg load signal;
end
% Menu principal
opcion = 0;
while opcion ~= 5
  % Menu de opciones
  disp('Seleccione una opcion:')
  disp('1. Grabar')
  disp('2. Reproducir')
  disp('3. Graficar')
  disp('4. Graficar densidad')
  disp('5. Salir')
  opcion = input('Ingrese su eleccién: ');
    switch opcion
      case 1
        %Grabacion de audio
        try
          duracion = input('Ingrese la duracién de la grabacién en segundos: ');
          disp('Comenzando la grabacién...');
          recObj = audiorecorder;
          recordblocking(recObj, duracion);
          disp('Grabacion finalizada.');
          data = getaudiodata(recObj);
          audiowrite('audio.wav', data, recObj.SampleRate);
          disp('Archivo de audio grabado correctamente.');
        catch
          disp('Error al grabar el audio.');
        end
      case 2
        % Reproduccién de audio
        try
          [data, fs] = audioread('audio.wav');
          sound(data, fs);
        catch
          disp('Error al reproducir el audio.');
        end
      case 3
        % Grafico de audio
        try
          [data, fs] = audioread('audio.wav');
          tiempo = linspace(0, length(data)/fs, length(data));
          plot(tiempo, data);
          xlabel('Tiempo (s)');
          ylabel('Amplitud');
          title('Audio');
        catch
          disp('Error al graficar el audio.');
        end
      case 4
        % Graficando espectro de frecuencia
      try
        disp('Graficando espectro de frecuencia...');
        [audio, Fs] = audioread('audio.wav'); % Lee la senal desde el archivo .wav
        N = length(audio); % Numero de muestras de la senal
        f = linspace(0, Fs/2, N/2+1); % Vector de frecuencias
        ventana = hann(N); %reducir el efecto de las discontinuidades al calcular la FFT
        Sxx = pwelch(audio, ventana, 0, N, Fs); % Densidad espectral de potencia
        plot(f, 10*log10(Sxx(1:N/2+1))); % Grafica el espectro de frecuencia en dB
        xlabel('Frecuencia (Hz)');
        ylabel('Densidad espectral de potencia (dB/Hz)');
        title('Espectro de frecuencia de la sefial grabada');
      catch
        disp('Error al graficar el audio.');
      end
      case 5
      % Salir
        disp('Saliendo del programa...');
      otherwise
        disp('Opcion no valida.');
    end
  end
